File: frontend/src/personalization/conceptIdentity.ts
/** Lexical identity only. Contextual meaning and taxonomy come from the observer. */
export function canonicalConceptName(value: string): string {
  let text = value.trim().replace(/^`|`$/g, "").replace(/\s+/g, " ");
  if (text.length < 2 || text.length > 120 || value.includes("\n")) return "";
  if (/[。！？!?；;，=\[\]{}"“”]|为什么|怎么|如何|下一步|请解释|它要求/.test(text)) return "";
  if (/[<>]/.test(text)) {
    let depth = 0;
    let output = "";
    for (const char of text) {
      if (char === "<") depth++;
      else if (char === ">") { if (--depth < 0) return ""; }
      else if (depth === 0) output += char;
    }
    if (depth || !/^[A-Za-z_]\w*(?:(?:::|\.)[A-Za-z_]\w*)*$/.test(output)) return "";
    text = output;
  }
  if (/[(),（）/\\]|^(?:return|delete|new|void\*)\b/.test(text)) return "";
  if ((text.match(/[\u4e00-\u9fff]/g)?.length ?? 0) > 16 || text.split(" ").length > 5) return "";
  return text;
}

export function mentionsConcept(text: string, name: string): boolean {
  if (!canonicalConceptName(name)) return false;
  const left = /^[a-zA-Z0-9_]/.test(name) ? "(?<![a-zA-Z0-9_])" : "";
  const right = /[a-zA-Z0-9_]$/.test(name) ? "(?![a-zA-Z0-9_])" : "";
  const escaped = name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  return new RegExp(left + escaped + right, "i").test(text);
}

File: backend/app/services/personalization/concept_identity.py
"""Conservative lexical identity. Semantic taxonomy belongs to the observer."""
from __future__ import annotations

import re


def canonical_concept_name(value: str) -> str:
    text = re.sub(r"\s+", " ", value.strip().strip("`"))
    if not 2 <= len(text) <= 120 or "\n" in value:
        return ""
    if re.search(r"[。！？!?；;，=\[\]{}\"“”]|(?:为什么|怎么|如何|下一步|请解释|它要求)", text):
        return ""
    # Generic arguments may contain commas; validate punctuation after stripping.
    # Retain member identity; never infer a namespace for a bare identifier.
    if "<" in text or ">" in text:
        depth = 0
        output = []
        for char in text:
            if char == "<":
                depth += 1
            elif char == ">":
                depth -= 1
                if depth < 0:
                    return ""
            elif depth == 0:
                output.append(char)
        if depth:
            return ""
        text = "".join(output)
        if not re.fullmatch(r"[A-Za-z_]\w*(?:(?:::|\.)[A-Za-z_]\w*)*", text):
            return ""
    if re.search(r"[(),（）/\\]|^(?:return|delete|new|void\*)\b", text):
        return ""
    if len(re.findall(r"[\u4e00-\u9fff]", text)) > 16 or len(text.split()) > 5:
        return ""
    return text


def mentions_concept(text: str, name: str) -> bool:
    if not canonical_concept_name(name):
        return False
    left = r"(?<![a-zA-Z0-9_])" if re.match(r"[a-zA-Z0-9_]", name) else ""
    right = r"(?![a-zA-Z0-9_])" if re.search(r"[a-zA-Z0-9_]$", name) else ""
    return re.search(left + re.escape(name) + right, text, re.IGNORECASE) is not None
